### DISCUSSION TITLE
When to add scheduling option .immediate?

### DISCUSSION THREAD

**Discussion by Jasperav** [groue/GRDBQuery#9] [Q&A] [answered]
This is from the documentation: 

`The scheduling: .immediate option should be removed for database requests that are too slow.`

Does this apply to a network request, or just a slow query? I want to show a spinner when a user sees my screen, because I am doing a network request to populate my database. After that, I want to show the list. 

Which scheduling option should I use in the case of a network request?

**Comment by groue** [comment, marked as answer]
The `scheduling: .immediate` option only applies to ValueObservation ([documentation](https://github.com/groue/GRDB.swift/blob/master/README.md#valueobservation-scheduling)). It has the database observation publisher publish its first element *immediately*, right on subscription.

The publisher is subscribed when a view that contains a `@Query` property enters the screen, when requested by SwiftUI, from the main thread. When this publisher is `.immediate`, this allows the view to display database value right from the start (with multiple benefits: see [Why `@Query`?](https://github.com/groue/GRDBQuery/blob/main/README.md#why-query)).

But when the database request is slow, this will block the main thread until the ValueObservation has performed its initial fetch. The user may see a noticeable delay before the view appears on screen, and the app won't be responsive.

That's why "database requests that are too slow" should not use the `scheduling: .immediate` option. Without this option, the initial fetch will happen in the background, and the main thread will not be blocked. But the view will have to display _something_ until the fetch completes, and that _something_ is the default value of the `Queryable` type.

> Does this apply to a network request, or just a slow query? I want to show a spinner when a user sees my screen, because I am doing a network request to populate my database. After that, I want to show the list.

If your `Queryable` type returns a publisher that can't publish anything until a network request has completed, then you can use the `defaultValue` in order to tell your view to display the spinner. And if the publisher starts observing the database after the network request has completed, then do not use the `.immediate` option, since *you do not need it*.

**Reply by Jasperav** [reply]
Alright thanks I got it, I will use the `async(onQueue)` display a spinner while the network request is going on

**Reply by groue** [reply]
Yes! Mind `async(onQueue: .main)` is the default (you don't have to specify anything).
